core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -2361,6 +2364,36 @@ public static int subtractMonths(long t0, long t1) {
     return x;
   }
 
+  /**
+   * Implements the {@code .} (field access) operator on an object whose type is not known until
+   * runtime.
+   *
+   * A struct object can be represented in various ways by the runtime and depends on the
+   * {@link org.apache.calcite.adapter.enumerable.JavaRowFormat}.
+   */
+  @Experimental
+  public static Object structAccess(Object structObject, int index, String fieldName) {
+    if (structObject == null) {
+      return null;
+    }
+
+    if (structObject instanceof Object[]) {
+      return ((Object[]) structObject)[index];
+    } else if (structObject instanceof List) {
+      return ((List) structObject).get(index);
+    } else if (structObject instanceof Row) {
+      return ((Row) structObject).getObject(index);
+    } else {
+      Class<?> beanClass = structObject.getClass();
+      try {
+        Field structField = beanClass.getDeclaredField(fieldName);
+        return structField.get(structObject);
+      } catch (NoSuchFieldException | IllegalAccessException ex) {
+        throw new IllegalStateException(ex);
```

Comment:
Would you please add relevant exception message?
That is `fieldName`/`index` that triggered the failure.

Reply:
I added fieldName but not index since the latter is not going to be used to raise this exception.